src/Serilog/Parameters/PropertyValueConverter.cs

```diff
@@ -154,17 +155,28 @@ LogEventPropertyValue CreatePropertyValue(object value, Destructuring destructur
                 // Only actual dictionaries are supported, as arbitrary types
                 // can implement multiple IDictionary interfaces and thus introduce
                 // multiple different interpretations.
+                var valueAsList = enumerable.Cast<object>();
                 if (IsValueTypeDictionary(valueType))
                 {
-                    return new DictionaryValue(enumerable.Cast<dynamic>()
-                        .Select(kvp => new KeyValuePair<ScalarValue, LogEventPropertyValue>(
-                                           (ScalarValue)limiter.CreatePropertyValue(kvp.Key, destructuring),
-                                           limiter.CreatePropertyValue(kvp.Value, destructuring)))
-                        .Where(kvp => kvp.Key.Value != null));
+                    var values = new List<KeyValuePair<ScalarValue, LogEventPropertyValue>>();
+                    if (valueAsList.Any())
+                    {
+                        var typeInfo = valueAsList.ElementAt(0).GetType().GetTypeInfo();
```

Comment:
Instead of `Any`/`ElementAt`, could we put the `keyProperty` and `valueProperty` members outside the `Select` block and lazily initialize them when the first item is processed?

(I'm guessing you've already considered it, but just in case you can think of a nicer way to do it than I, the value type will always be a predictable one: `typeof(KeyValuePair<,>).MakeGenericType(valueType.GetGenericArguments())`. Not sure about the perf of this approach, but would save having to examine any elements.)